include/deal.II/differentiation/ad/ad_helpers.h

```diff
@@ -0,0 +1,683 @@
+     * variables in different ways, above represents only an outline of the
+     * steps taken to compute derivatives. More specific examples are outlined
+     * in the individual classes that specialize this base class.
+     *
+     * When "taped" numbers are to be used, the above code should be wrapped by
+     * a few more lines of code to manage the taping procedure:
+     *
+     * @code
+     *   // Initialize AD helper
+     *   ADHelperType<dim,taped_or_tapeless_AD_typecode> ad_helper (...);
+     *
+     *   // An optional call to set the amount of memory to be allocated to
+     *   // storing taped data
+     *   ad_helper.set_tape_buffer_sizes();
+     *
+     *    // Select a tape number to record to
+     *   const types::tape_index tape_index = ...;
+     *
+     *   // Indicate that we are about to start tracing the operations for
+     *   // function evaluation on the tape. If this tape has already been used
+     *   // (i.e. the operations are already recorded) then we (optionally) load
+     *   // the tape and reuse this data.
+     *   const bool is_recording
+     *     = ad_helper.enable_record_sensitivities(tape_index);
+     *   if (is_recording == true)
+     *   {
+     *     // In this block on places the majority of the operations described
+     *     // in the previous code block. The set of operations that are
+     *     // conducted here therefore includes the following steps:
+     *     // - Register independent variables
+     *     // - Extract the sensitive equivalent of the independent variables
+     *     // - Use the sensitive variables to compute the dependent variables
+     *     // - Register the dependent variables with the helper class
+     *   }
+     *   else
+     *   {
+     *     // Load the existing tape to be reused
+     *     ad_helper.activate_recorded_tape(tape_no);
+     *
+     *     // Set the new values of the independent variables where the recorded
+     *     // dependent functions are to be evaluated (and differentiated
+     *     // around).
+     *     ad_helper.set_independent_variable(...);
+     *   }
+     *
+     *   // Compute derivatives of the dependent variables
+     *   const auto derivatives = ad_helper.compute_gradients();
+     * @endcode
+     *
+     * The second approach outlined here is more general than the first, and
+     * will work equally well for both taped and tapeless auto-differentiable
+     * numbers.
+     *
+     * @warning ADOL-C does not support the standard threading models used by
+     * deal.II, so this class should \b not be embedded within a multithreaded
+     * function when using ADOL-C number types. It is, however, suitable for use
+     * in both serial and MPI routines.
+     *
+     * @author Jean-Paul Pelteret, 2016, 2017, 2018
+     */
+    template <int                  dim,
+              enum AD::NumberTypes ADNumberTypeCode,
+              typename ScalarType = double>
+    class ADHelperBase
```

Comment:
I can't seem to figure out where `dim` is actually used in these classes. Can you elaborate?